File: backend/app/models/invoice.py
from sqlalchemy import Column, String, Float, Boolean, DateTime, Text, ForeignKey, Integer
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.database import Base
import uuid

class Invoice(Base):
    __tablename__ = "invoices"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    owner_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False)

    # Client details
    client_name = Column(String(200), nullable=False)
    client_email = Column(String(255), nullable=True)
    client_phone = Column(String(20), nullable=True)
    client_address = Column(Text, nullable=True)

    # Invoice details
    invoice_number = Column(String(50), unique=True, nullable=False)
    invoice_date = Column(DateTime(timezone=True), server_default=func.now())
    due_date = Column(DateTime(timezone=True), nullable=True)
    event_name = Column(String(200), nullable=True)
    event_date = Column(DateTime(timezone=True), nullable=True)

    # Line items stored as JSON
    items = Column(Text, nullable=False)  # JSON list of {description, quantity, rate, amount}

    # Amounts
    subtotal = Column(Float, nullable=False, default=0)
    tax_percent = Column(Float, nullable=False, default=18.0)  # GST
    tax_amount = Column(Float, nullable=False, default=0)
    discount_amount = Column(Float, nullable=False, default=0)
    total_amount = Column(Float, nullable=False, default=0)

    # Payment
    status = Column(String(20), default="pending")  # pending, paid, overdue, cancelled
    razorpay_payment_link_id = Column(String(100), nullable=True)
    razorpay_payment_link_url = Column(Text, nullable=True)
    razorpay_payment_id = Column(String(100), nullable=True)
    paid_at = Column(DateTime(timezone=True), nullable=True)

    # Notes
    notes = Column(Text, nullable=True)

    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    owner = relationship("User", backref="invoices")

File: frontend/src/lib/invoice.ts
import api from "./api";

export interface InvoiceItem {
    description: string;
    quantity: number;
    rate: number;
    amount: number;
}

export interface Invoice {
    id: string;
    invoice_number: string;
    client_name: string;
    client_email: string | null;
    client_phone: string | null;
    event_name: string | null;
    subtotal: number;
    tax_amount: number;
    discount_amount: number;
    total_amount: number;
    tax_percent: number;
    status: "pending" | "paid" | "overdue" | "cancelled";
    razorpay_payment_link_url: string | null;
    invoice_date: string;
    due_date: string | null;
    paid_at: string | null;
    notes: string | null;
    items: string;
    created_at: string;
}

export interface Analytics {
    monthly: {
        total_billed: number;
        collected: number;
        pending: number;
        invoice_count: number;
    };
    all_time: {
        total_billed: number;
        collected: number;
        pending: number;
        invoice_count: number;
    };
    by_status: {
        pending: number;
        paid: number;
        overdue: number;
    };
}

export interface StudioProfile {
    name: string;
    email: string;
    studio_name: string | null;
    studio_address: string | null;
    studio_phone: string | null;
    studio_gstin: string | null;
    studio_upi_id: string | null;
}

export const invoiceService = {
    async list(): Promise<Invoice[]> {
        const res = await api.get("/invoices");
        return res.data;
    },

    async get(id: string): Promise<Invoice> {
        const res = await api.get(`/invoices/${id}`);
        return res.data;
    },

    async create(data: {
        client_name: string;
        client_email?: string;
        client_phone?: string;
        client_address?: string;
        event_name?: string;
        event_date?: string;
        due_date?: string;
        items: InvoiceItem[];
        tax_percent: number;
        discount_amount: number;
        notes?: string;
    }): Promise<Invoice> {
        const res = await api.post("/invoices", data);
        return res.data;
    },

    async markPaid(id: string): Promise<Invoice> {
        const res = await api.patch(`/invoices/${id}/mark-paid`);
        return res.data;
    },

    async cancel(id: string): Promise<Invoice> {
        const res = await api.patch(`/invoices/${id}/cancel`);
        return res.data;
    },

    async delete(id: string): Promise<void> {
        await api.delete(`/invoices/${id}`);
    },

    async sendEmail(id: string): Promise<void> {
        await api.post(`/invoices/${id}/send-email`);
    },

    async downloadPdf(id: string, invoiceNumber: string): Promise<void> {
        const res = await api.get(`/invoices/${id}/pdf`, { responseType: "blob" });
        const url = URL.createObjectURL(res.data);
        const a = document.createElement("a");
        a.href = url;
        a.download = `Invoice_${invoiceNumber}.pdf`;
        a.click();
        URL.revokeObjectURL(url);
    },

    async getAnalytics(): Promise<Analytics> {
        const res = await api.get("/invoices/analytics");
        return res.data;
    },

    async getProfile(): Promise<StudioProfile> {
        const res = await api.get("/invoices/profile");
        return res.data;
    },

    async updateProfile(data: Partial<StudioProfile>): Promise<void> {
        await api.put("/invoices/profile", data);
    },
};